tests/test_cli.py

```diff
@@ -25,73 +25,74 @@ def cleanoutputdir() -> Generator[None, None, None]:
 
 
 @pytest.fixture
-def options(tmp_path: Path) -> List[str]:
-    return [
-        "--template",
-        "tests/data/example.docx",
-        "--data",
-        "tests/data/example.csv",
-        "--name",
-        "NAME",
-        "--path",
-        str(tmp_path.resolve()),
-    ]
+def options_gen(tmp_path: Path) -> Callable:
+    def _options_gen(**kwargs: str) -> List:
+        default = {
+            "template": "tests/data/example.docx",
+            "data": "tests/data/example.csv",
+            "name": "NAME",
+            "delimiter": ";",
+            "path": str(tmp_path.resolve()),
+        }
 
+        # override values if provided
+        for key, value in kwargs.items():
+            if key in default:
+                default[key] = value
 
-@pytest.mark.usefixtures("cleanoutputdir")
-def test_basic_succeeds(runner: CliRunner, options: List) -> None:
-
-    del options[-2:]
-
-    result = runner.invoke(cli.main, options)
+        # convert dict to sequential list, add '--' for CLI options (i.e. the keys)
+        return [x for (key, value) in default.items() for x in ("--" + key, value)]
 
-    assert result.exit_code == 0
+    return _options_gen
 
 
-def test_custom_output_path_succeeds(runner: CliRunner, options: List) -> None:
-    """testing pathlib.Path object as output"""
+@pytest.mark.usefixtures("cleanoutputdir")
+def test_defaults(runner: CliRunner, options_gen: Callable) -> None:
 
-    result = runner.invoke(cli.main, options)
+    result = runner.invoke(cli.main, options_gen()[:-2])
 
     assert result.exit_code == 0
 
 
-def test_custom_deep_output_path_succeeds(runner: CliRunner, options: List) -> None:
-    """testing deep non-existent directory"""
-
-    options[-1] = Path(options[-1]) / "deep" / "sub" / "folder"
+def test_output_path(runner: CliRunner, options_gen: Callable) -> None:
 
-    result = runner.invoke(cli.main, options)
+    result = runner.invoke(cli.main, options_gen())
 
     assert result.exit_code == 0
 
 
-def test_custom_output_dir_succeeds(
-    runner: CliRunner, options: List, tmpdir: Path
+def test_deep_output_path(
+    runner: CliRunner, options_gen: Callable, tmp_path: Path
 ) -> None:
-    """testing py.path.local object as output
-    the `tmpdir` provides a temporary directory unique to this test invocation"""
 
-    options[-1] = tmpdir
-
-    result = runner.invoke(cli.main, options)
+    result = runner.invoke(
+        cli.main, options_gen(path=tmp_path / "deep" / "sub" / "folder")
+    )
 
     assert result.exit_code == 0
 
 
-def test_output_name_not_found(runner: CliRunner, options: List) -> None:
+def test_invalid_file(runner: CliRunner, options_gen: Callable) -> None:
+
+    result = runner.invoke(cli.main, options_gen(data="125123ssdfv9a98a7e43"))
 
-    options[5] = "WRONG_NAME"
+    assert result.exit_code == 0
+    assert "could not be opened or found." in result.output
```

Comment:
@davidverweij - I wonder if instead of testing the string output we should raise and test exceptions instead, i.e., for failure?

Here we see that `result.exit_code` is 0, which indicates success? Whereas given it is failing, we should test against 1?

If we use `click.ClickException` as above, then perhaps we could do:


```python
    assert result.exit_code == 1
    assert result.exception
    # if you really want string comparisons
    assert "could not be opened or found." in result.output
```